### PR TITLE
Pt 166731397 add compiler backend

### PR DESCRIPTION
**Pull request by hanssv**

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -11,7 +11,7 @@
         {aeserialization, {git, "https://github.com/aeternity/aeserialization.git",
                            {ref, "3416ff5"}}},
         {aesophia, {git, "https://github.com/aeternity/aesophia.git",
-                    {tag, "v3.2.0"}}},
+                    {ref, "6a27c4a"}}},
```
Comment (ThomasArts) [inline rebar.config]: That's not the newest

Comment (hanssv) [reply rebar.config]: It is new enough for this application :-) We will sync them properly when it is time for release.

File: apps/aesophia_http/src/aesophia_http_handler.erl [hunk 2]
```diff
@@ -201,13 +210,19 @@ compile_options(Options) ->
     Map = maps:get(<<"file_system">>, Options, #{}),
     Map1 = maps:from_list([{binary_to_list(N), F} || {N, F} <- maps:to_list(Map)]),
     SrcFile = maps:get(<<"src_file">>, Options, no_file),
-    [{include, {explicit_files, Map1}}]
+    Backend = case binary_to_atom(maps:get(<<"backend">>, Options, <<"default">>), utf8) of
+                  aevm -> aevm;
+                  _    -> fate
```
Comment (UlfNorell) [inline apps/aesophia_http/src/aesophia_http_handler.erl]: Might want to only allow `fate` and `default` here, and fail for `aeevm` (say).

File: apps/aesophia_http/src/aesophia_http_handler.erl [hunk 3]
```diff
@@ -253,15 +268,22 @@ parse_type(BinaryString) ->
...
-        {ok, #{type_info := TypeInfo}} ->
-            decode_calldata_bytecode_(Calldata, TypeInfo);
+        %% Try a bit to be clever, if backend is not set - do some
+        %% rudimentary auto detection.
+        {ok, #{type_info := [], byte_code := Bytecode}} when Backend == default ->
+            decode_calldata_bytecode_(fate, Calldata, Bytecode);
+        {ok, #{type_info := TypeInfo}} when Backend == default; Backend == aevm ->
+            decode_calldata_bytecode_(aevm, Calldata, TypeInfo);
+        {ok, #{byte_code := Bytecode}} ->
+            decode_calldata_bytecode_(fate, Calldata, Bytecode);
```
Comment (UlfNorell) [inline apps/aesophia_http/src/aesophia_http_handler.erl]: Same here. Better to fail on unknown backend than to default to fate.

File: apps/aesophia_http/src/aesophia_http_handler.erl [hunk 4]
```diff
@@ -280,8 +302,30 @@ decode_calldata_bytecode_(Calldata, TypeInfo) ->
...
+        {403, [], #{reason => <<"Could not deserialize FATE bytecode">>}}
     end.
 
+decode_calldata_fatecode(FunHash, Args, FCode) ->
+    case aeb_fate_abi:get_function_name_from_function_hash(FunHash, FCode) of
+        {ok, FunName} ->
+            {200, [], #{function => FunName,
+                        arguments => [fate_to_json(Arg) || Arg <- Args]}};
+        _ ->
+            {403, [], #{reason => <<"Could not find function hash in FCode">>}}
```
Comment (UlfNorell) [inline apps/aesophia_http/src/aesophia_http_handler.erl]:
```suggestion
            {403, [], #{reason => <<"Could not find function hash in FATE bytecode">>}}
```

File: apps/aesophia_http/src/aesophia_http_handler.erl [hunk 5]
```diff
@@ -430,3 +476,23 @@ prepare_for_json(T, R) ->
...
+
+jo(T, V) -> #{ <<"type">> => atom_to_binary(T, utf8), <<"value">> => V }.
+
+fate_to_json(?FATE_ADDRESS(Bin))  -> jo(address, aeser_api_encoder:encode(account_pubkey, Bin));
+fate_to_json(?FATE_ORACLE(Bin))   -> jo(oracle, aeser_api_encoder:encode(oracle_pubkey, Bin));
+fate_to_json(?FATE_ORACLE_Q(Bin)) -> jo(oracle_query, aeser_api_encoder:encode(oracle_query_id, Bin));
+fate_to_json(?FATE_CONTRACT(Bin)) -> jo(contract, aeser_api_encoder:encode(contract_pubkey, Bin));
+fate_to_json(?FATE_BYTES(Bin))    -> jo(bytes, aeser_api_encoder:encode(bytearray, Bin));
+fate_to_json(?FATE_BITS(Bin))     -> jo(bits, aeser_api_encoder:encode(bytearray, Bin));
+fate_to_json(N) when is_integer(N) -> jo(int, N);
```
Comment (UlfNorell) [inline apps/aesophia_http/src/aesophia_http_handler.erl]: Should really be
```suggestion
fate_to_json(N) when ?IS_FATE_INTEGER(N) -> jo(int, ?FATE_INTEGER_VALUE(N));
```
Similar for bool, string, list and map below.

**Pull request merged**